### DISCUSSION TITLE
Multiple kinematic hardening models

### DISCUSSION THREAD

**Discussion by asingh1226** [thelfer/tfel#189] [General]
We have been using mfront/tfel/fenics for plasticity models and the related standard elastic-visco-plasticity brick:

http://tfel.sourceforge.net/StandardElastoViscoPlasticityBrick.html [Ref 1]

Here we found that multiple isotropic hardening models can be used. For example,  see Section 2.4 in Ref 1, where two Voce isotropic hardening models have been used:

   isotropic_hardening : "Voce" {R0 : 600e6, Rinf : 900e6, b : 1},
   isotropic_hardening : "Voce" {R0 : 0, Rinf : 300e6, b : 10},

However, we were not able to implement multiple kinematic hardening models in the similar manner. For example, suppose we would like to implement two backstress models in Chaboche 2012, but as of now only one set of C, D, n and w parameters can be used, e.g., 
 kinematic_hardening : "Chaboche 2012" {
      C : 250e7,
      D : 100,
      m : 2,
      w : 0.6,
    }

We thought that another set could be used by simply repeating the call by adding another "Chaboche 2012" in the following sense:
 kinematic_hardening : "Chaboche 2012" {
      C : 250e7,
      D : 100,
      m : 2,
      w : 0.6,
    }
 kinematic_hardening : "Chaboche 2012" {
      C : 3e11,
      D : 500,
      m : 1,
      w : 0.0,
    }

Is there a way it can be done by changing the code?

Another option, of course, is to write another kinematic hardening model which has the parameters from both models. Any suggestions?

**Comment by asingh1226** [comment]
Just to clarify, the back stress in Chaboche can be decomposed as a sum of several independent variables. At present, mfront StandardElastoViscoPlasticityBrick is allowing only one back stress variable. See Eq. 5 in https://www.sciencedirect.com/science/article/pii/S0749641912000113

**Comment by thelfer** [comment]
Hi @asingh1226, I am quite surprised. The following snipset compiles:

```cxx
@DSL Implicit;
@Behaviour Test;
@Author Thomas Helfer;
@Date 13 / 05 / 2022;

@Brick StandardElastoViscoPlasticity{
  stress_potential : "Hooke" {young_modulus : 150e9, poisson_ratio : 0.3},
  inelastic_flow : "Plastic" {
    criterion : "Mises",
    isotropic_hardening : "Linear" {R0 : 150e6, H : 2e9},
    kinematic_hardening : "Chaboche 2012" {C : 250e7, D : 100, m : 2, w : 0.6},
    kinematic_hardening : "Chaboche 2012" {C : 3e11, D : 500, m : 1, w : 0.0}
  }
};
```

Would you try ?
Thomas

**Comment by asingh1226** [comment]
Surprising. So it looks like we have a different problem. Yours script is definitely compiling. But when I use more than one isotropic hardening models, its not compiling. Could you please verify this:

@DSL Implicit;
 @Behaviour Test;
 @Author Thomas Helfer;
 @Date 13 / 05 / 2022;
 
 @Brick StandardElastoViscoPlasticity{
   stress_potential : "Hooke" {young_modulus : 150e9, poisson_ratio : 0.3},
   inelastic_flow : "Plastic" {
     criterion : "Mises",
     isotropic_hardening : "Linear" {R0 : 150e6, H : 2e9},
     isotropic_hardening : "Voce" {R0 : 260e6, Rinf : 260e6, b : 31.51},
     kinematic_hardening : "Chaboche 2012" {C : 250e7, D : 100, m : 2, w : 0.6},
     kinematic_hardening : "Chaboche 2012" {C : 3e11, D : 500, m : 1, w : 0.0}
   }
 };

**Reply by thelfer** [reply]
@asingh1226 Indeed, I got this error:

```bash
BehaviourDSLCommon::analyse: error while treating keyword '@Brick' at line '6' of file '/tmp/example.mfront'.
GenTypeCastError: bad cast
```

which disapears if I only declare one isotropic hardening rule. Seems a silly bug. Let me have a look at it.

**Comment by asingh1226** [comment]
Thanks for looking into it. Would be great if we can fix the bug.

**Reply by thelfer** [reply]
Just closed Issue #190. Shall compile now. Happy testing !
